File: modules/cms/cms-internal/src/main/java/org/eclipse/dirigible/cms/internal/Document.java
/*
 * Copyright (c) 2022 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.dirigible.cms.internal;

import java.io.ByteArrayInputStream;
import java.io.IOException;

import org.eclipse.dirigible.repository.api.IRepository;
import org.eclipse.dirigible.repository.api.IResource;

/**
 * The Class Document.
 */
public class Document extends CmisObject {

	/** The session. */
	private CmisSession session;

	/** The internal resource. */
	private IResource internalResource;

	/** The repository. */
	private IRepository repository;

	/**
	 * Instantiates a new document.
	 *
	 * @param session the session
	 * @param internalResource the internal resource
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	public Document(CmisSession session, IResource internalResource) throws IOException {
		super(session, internalResource.getPath());
		this.session = session;
		this.repository = (IRepository) session.getCmisRepository().getInternalObject();
		this.internalResource = internalResource;
	}

	/**
	 * Instantiates a new document.
	 *
	 * @param session the session
	 * @param id the id
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	public Document(CmisSession session, String id) throws IOException {
		super(session, id);
		id = sanitize(id);
		this.session = session;
		this.repository = (IRepository) session.getCmisRepository().getInternalObject();
		this.internalResource = this.repository.getResource(id);
	}

	/**
	 * Gets the internal folder.
	 *
	 * @return the internal folder
	 */
	public IResource getInternalFolder() {
		return internalResource;
	}

	/**
	 * Checks if is collection.
	 *
	 * @return true, if is collection
	 */
	@Override
	protected boolean isCollection() {
		return false;
	}

	/**
	 * Returns the ContentStream representing the contents of this Document.
	 *
	 * @return Content Stream
	 * @throws IOException IO Exception
	 */
	public ContentStream getContentStream() throws IOException {
		byte[] content = this.internalResource.getContent();
		return new ContentStream(session, this.internalResource.getName(), content.length, this.internalResource.getContentType(),
				new ByteArrayInputStream(content));
	}

	/**
	 * Returns the Path of this Document.
	 *
	 * @return the path
	 */
	public String getPath() {
		return this.internalResource.getPath();
	}

}

File: components/ide/ide-ui-documents/src/main/resources/META-INF/dirigible/ide-documents/utils/cmis/document.js
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
import { cmis } from "sdk/cms";
import { streams } from "sdk/io";
import * as objectUtils from "./object";
const path = org.eclipse.dirigible.repository.api.RepositoryPath;

let cmisSession = cmis.getSession();

function DocumentSerializer(cmisDocument) {
	this.id = cmisDocument.getId();
	this.name = cmisDocument.getName();
}

export const uploadDocument = (folder, document) => {
	let fileName = getDocumentFileName(document);
	let mimetype = document.getContentType();
	let size = document.getSize();
	let inputStream = document.getInputStream();
	let newDocument = createDocument(folder, fileName, size, mimetype, inputStream);
	return new DocumentSerializer(newDocument);
};

export const uploadDocumentOverwrite = (folder, document) => {
	let timestamp = new Date().getTime();
	if (document.name === null || document.name === undefined) {
		document.name = document.getName();
	}
	let newName = document.name + "-" + timestamp;
	let oldName = document.name;

	document.name = newName;
	uploadDocument(folder, document);

	try {
		let docPath = path.normalizePath(folder.getPath(), oldName)
		let oldDoc = objectUtils.getObject(docPath);
		objectUtils.deleteObject(oldDoc);
	} catch (e) {
		//do nothing
	}
	let docPath = path.normalizePath(folder.getPath(), newName)
	let newDoc = objectUtils.getObject(docPath);
	objectUtils.renameObject(newDoc, oldName);
};

function getDocumentFileName(document) {
	//check for .name first as it's passed from uploadDocumentOverwrite
	return document.name ?? document.getName();
}

function createDocument(folder, fileName, size, mimetype, inputStream) {
	let contentStream = cmisSession.getObjectFactory().createContentStream(fileName, size, mimetype, inputStream);
	let properties = {};
	properties[cmis.OBJECT_TYPE_ID] = cmis.OBJECT_TYPE_DOCUMENT;
	properties[cmis.NAME] = fileName;
	let newDocument = folder.createDocument(properties, contentStream, cmis.VERSIONING_STATE_MAJOR);
	return newDocument;
}

export const getDocumentStream = (document) => {
	let contentStream = document.getContentStream();
	return contentStream;
};

export const getDocNameAndStream = (document) => {
	let stream = getDocumentStream(document);
	let name = document.getName();
	return [name, stream];
};

export const getDocument = (path) => {
	return objectUtils.getObject(path);
}

export const existDocument = (path) => {
	return objectUtils.existObject(path);
}

export const createFromBytes = (folder, fileName, bytes) => {
	let inputStream = streams.createByteArrayInputStream(bytes);
	let mimeType = "application/octet-stream";

	createDocument(folder, fileName, bytes.length, mimeType, inputStream);
};